libherokubuildpack/src/output/background.rs

```diff
@@ -0,0 +1,213 @@
+
+        /// Holds the signaling method to tell the background printer
+        /// to stop emitting.
+        stop_signal: Sender<()>,
+    }
+
+    impl<W> Drop for PrintGuard<W>
+    where
+        W: Write + Debug,
+    {
+        fn drop(&mut self) {
+            // A note on correctness. It might seem that it's enough to signal the thread to
+            // stop, that we don't also have to join and wait for it to finish, but that's not
+            // the case. The printer can still emit a value after it's told to stop.
+            //
+            // When that happens the output can appear in the middle of another output, such
+            // as an error message if a global writer is being used such as stdout.
+            // As a result we have to signal AND ensure the thread is stoped before
+            // continuing.
+            if let Some(join_handle) = self.join_handle.take() {
+                let _ = self.stop_signal.send(());
+                let _ = join_handle.join();
+            }
+        }
+    }
+
+    impl<W> PrintGuard<W>
+    where
+        W: Write + Debug,
+    {
+        /// Preserve internal state by ensuring the `Option` is always populated
+        pub(crate) fn new(join_handle: JoinHandle<W>, sender: Sender<()>) -> Self {
+            let guard = PrintGuard {
+                join_handle: Some(join_handle),
+                stop_signal: sender,
+            };
+            debug_assert!(guard.join_handle.is_some());
+
+            guard
+        }
+
+        /// The only thing a consumer can do is stop the dots and receive
+        /// the original buffer.
+        ///
+        /// # Panics
+        ///
+        /// This code can panic if it encounters an unexpedcted internal state.
+        /// If that happens it means there is an internal bug in this logic.
+        /// To avoid a panic, developers modifying this file must obey the following
+        /// rules:
+        ///
+        /// - Always consume the struct when accessing the Option value outside of Drop
+        /// - Never construct this struct with a `None` option value.
+        ///
+        /// This `Option` wrapping is needed to support a implementing Drop to
+        /// ensure the printing is stopped. When a struct implements drop it cannot
+        /// remove it's internal state due to E0509:
+        ///
+        /// <https://github.com/rust-lang/rust/blob/27d8a577138c0d319a572cd1a464c2b755e577de/compiler/rustc_error_codes/src/error_codes/E0509.md>
+        ///
+        /// The workaround is to never allow invalid internal state by replacing the
+        /// inner value with a `None` when removing it. We don't want to expose this
+        /// implementation detail to the user, so instead we accept the panic, ensure
+        /// the code is exercised under test, and exhaustively document why this panic
+        /// exists and how developers working with this code can maintain safety.
+        #[allow(clippy::panic_in_result_fn)]
```

Comment:
(Redundant if we remove `print_interval` / background threads)

I think the `stop()` function can be refactored using combinators and `.expect()` to be clearer and also avoid the need for this `allow` (since it won't be using `panic!` any more)